crates/web-sys/webidls/enabled/MathMLElement.webidl

```diff
@@ -0,0 +1,19 @@
+/* -*- Mode: IDL; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * The origin of this IDL file is
+ * https://mathml-refresh.github.io/mathml-core/
+ *
+ * Copyright © 2019 W3C® (MIT, ERCIM, Keio, Beihang). W3C liability, trademark
+ * and permissive document license rules apply.
+ */
+
+[Exposed=Window]
+interface MathMLElement : Element { };
+MathMLElement includes GlobalEventHandlers;
+MathMLElement includes HTMLOrForeignElement;
+MathMLElement includes ElementCSSInlineStyle;
+MathMLElement includes TouchEventHandlers;
+MathMLElement includes OnErrorEventHandlerForNodes;
```

Comment:
I mentioned this before, but we follow the specification nowadays instead of browser implementations.
```idl
  interface MathMLElement : Element
  {
             attribute DOMString         className;
             attribute DOMString         mathElementStyle;
             attribute DOMString         id;
             attribute DOMString         xref;
             attribute DOMString         href;
    readonly attribute MathMLMathElement ownerMathElement;
  };
```

I found `MathMLElement includes ElementCSSInlineStyle` in [CSSOM](https://drafts.csswg.org/cssom/#ref-for-elementcssinlinestyle%E2%91%A2), each one of those needs a reference to the spec it comes from.

Reply:
I mean I can certainly add those attributes, but I'm not sure what the target group is, when it's not browser implementations (i.e. what other wasm target supports something like MathML?).

The issue I see with adding something like `mathElementStyle` is that there's no to me known wasm implementation that supports it, and it may be tempting to use, but results in exceptions then, I'm not sure if that's really what we want?

IME the spec(s) can be kinda messy and not really represent the reality (probably grown historically), that's why I think an intersection between being compliant to the specs and support of actual implementations is a good compromise.

I think [this spec](https://www.w3.org/TR/mathml-core/#dom-and-javascript) is more up-to-date and better represents the reality (funnily they mention `HTMLOrForeignElement` there and link to the `HTMLOrSVGElement`...)

> each one of those needs a reference to the spec it comes from.

I mean each one of those (mixins) has a link that specifies this indirectly in the mixin (but the `HTMLOrSVGElement` as you noticed, and I added a link to the `TouchEventHandlers`), should I do anymore here?

Reply:
> I mean I can certainly add those attributes, but I'm not sure what the target group is, when it's not browser implementations (i.e. what other wasm target supports something like MathML?).

No that's fine.
I was just trying to point out that all the `includes` aren't mentioned there, but I can see how this was confusing!

> > each one of those needs a reference to the spec it comes from.
> 
> I mean each one of those (mixins) has a link that specifies this indirectly in the mixin (but the `HTMLOrSVGElement` as you noticed, and I added a link to the `TouchEventHandlers`), should I do anymore here?

Basically I just want a URL next to each one of these `includes` to the spec that tells us where we got that from.
E.g. the link you added to `TouchEventHandlers` is great, but I can't find the line `MathMLElement includes TouchEventHandlers;` in it.

Reply:
>  but I can't find the line MathMLElement includes TouchEventHandlers; in it.

It's kind of implied in:

> The following section describes extensions to the existing [GlobalEventHandlers](https://www.w3.org/TR/html5/#globaleventhandlers) mixin, defined in [[HTML5](https://w3c.github.io/touch-events/#bib-html5)], to facilitate the event handler registration. For user agents where [expose legacy touch event APIs](https://w3c.github.io/touch-events/#dfn-expose-legacy-touch-event-apis) is false, this mixin must not be implemented.

```
partial interface mixin [GlobalEventHandlers](https://w3c.github.io/touch-events/#dom-globaleventhandlers) {
                    attribute [EventHandler](https://html.spec.whatwg.org/multipage/webappapis.html#eventhandler) [ontouchstart](https://w3c.github.io/touch-events/#dom-globaleventhandlers-ontouchstart);
                    attribute [EventHandler](https://html.spec.whatwg.org/multipage/webappapis.html#eventhandler) [ontouchend](https://w3c.github.io/touch-events/#dom-globaleventhandlers-ontouchend);
                    attribute [EventHandler](https://html.spec.whatwg.org/multipage/webappapis.html#eventhandler) [ontouchmove](https://w3c.github.io/touch-events/#dom-globaleventhandlers-ontouchmove);
                    attribute [EventHandler](https://html.spec.whatwg.org/multipage/webappapis.html#eventhandler) [ontouchcancel](https://w3c.github.io/touch-events/#dom-globaleventhandlers-ontouchcancel);
};
```

So to be more close to the spec, `TouchEventHandlers` should just be the above partial mixin `GlobalEventHandlers`.

I've added some additional comments, as I don't think it hurts.

Reply:
> It's kind of implied in:
> 
> > The following section describes extensions to the existing [GlobalEventHandlers](https://www.w3.org/TR/html5/#globaleventhandlers) mixin, defined in [[HTML5](https://w3c.github.io/touch-events/#bib-html5)], to facilitate the event handler registration. For user agents where [expose legacy touch event APIs](https://w3c.github.io/touch-events/#dfn-expose-legacy-touch-event-apis) is false, this mixin must not be implemented.
> 
> ```
> partial interface mixin GlobalEventHandlers {
>     attribute EventHandler ontouchstart;
>     attribute EventHandler ontouchend;
>     attribute EventHandler ontouchmove;
>     attribute EventHandler ontouchcancel;
> };
> ```
> 
> So to be more close to the spec, `TouchEventHandlers` should just be the above partial mixin `GlobalEventHandlers`.

I'm not following, if this should be `GlobalEventHandlers` instead, then why are we using `TouchEventHandlers` instead?
But I can't find any `MathMLElement includes GlobalEventHandlers` in any specification, so I'm not sure what you mean with implied.

I understand that this is just copied from the Mozilla source, but as I said, we don't do this anymore.
We only take from the specification.
> I've added some additional comments, as I don't think it hurts.

To clarify: this isn't optional, its the *requirement* I'm asking for here!

---

My ask here is that we don't add any WebIDL bindings outside from specifications.
Any added binding should contain a link to the specification its taken from.
If it doesn't exist in the specification, we don't add it.

Please let me know if anything is unclear!

Reply:
> I'm not following, if this should be GlobalEventHandlers instead, then why are we using TouchEventHandlers instead?

It's just the same as in `SvgElement` and `HtmlElement`, they have the same specs (`partial mixin GlobalEventHandlers`). We can of course also refactor these in the process as well, by using the partial mixin instead of the non-partial mixin `TouchEventHandlers`. The effect should be the same though.

> But I can't find any MathMLElement includes GlobalEventHandlers in any specification, so I'm not sure what you mean with implied.

Uh? I linked it in my previous comment (which is also in the (first) comment of the webidl code of the `MathMLElement`):

> I think [this spec](https://www.w3.org/TR/mathml-core/#dom-and-javascript) is more up-to-date and better represents the reality (funnily they mention `HTMLOrForeignElement` there and link to the `HTMLOrSVGElement`...)

> To clarify: this isn't optional, its the requirement I'm asking for here!

Right... but I don't think anything is missing anymore, i.e. all `includes` have a link to the relevant spec that defines it?

Reply:
Ah, I see now!

> Right... but I don't think anything is missing anymore, i.e. all `includes` have a link to the relevant spec that defines it (apart of `OnErrorEventHandlerForNodes` which has a comment in `EventHandler`, not sure about that)?

I will leave a separate comment on that.

Reply:
I guess you assumed I know whats going on with `TouchEventHandlers` and `OnErrorEventHandlerForNodes`, but unfortunately I didn't.

Sorry for the unnecessary back and forth!
Should be good to go after we resolve my suggestions below!